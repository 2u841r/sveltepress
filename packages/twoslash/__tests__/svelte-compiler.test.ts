import { describe, expect, it } from 'vitest'
import { parse, preprocess } from 'svelte/compiler'
import { vitePreprocess } from '@sveltejs/vite-plugin-svelte'

const code = `
<script context="module" lang="ts">
  export const FOO = 'BAR'
</script>
<script lang="ts">
  import { onMount } from 'svelte'

  let count = 0

  onMount(() => {
    console.log('mount')
  })
</script>
<button on:click="{count++}">
  Count is: { count }
</button>
`

describe('svelte compiler', () => {
  it('preprocess', async () => {
    expect(await preprocess(code, vitePreprocess())).toMatchSnapshot()
  })

  it('parse', async () => {
    expect(parse(code)).toMatchSnapshot()
  })
})
