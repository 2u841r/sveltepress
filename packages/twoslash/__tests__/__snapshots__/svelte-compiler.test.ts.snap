// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`svelte compiler > parse 1`] = `
{
  "css": undefined,
  "html": {
    "children": [
      {
        "data": "
",
        "end": 1,
        "raw": "
",
        "start": 0,
        "type": "Text",
      },
      {
        "data": "
",
        "end": 74,
        "raw": "
",
        "start": 73,
        "type": "Text",
      },
      {
        "data": "
",
        "end": 204,
        "raw": "
",
        "start": 203,
        "type": "Text",
      },
      {
        "attributes": [
          {
            "end": 232,
            "expression": Node {
              "argument": Node {
                "end": 228,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 24,
                    "line": 14,
                  },
                  "start": Position {
                    "column": 19,
                    "line": 14,
                  },
                },
                "name": "count",
                "start": 223,
                "type": "Identifier",
              },
              "end": 230,
              "loc": SourceLocation {
                "end": Position {
                  "column": 26,
                  "line": 14,
                },
                "start": Position {
                  "column": 19,
                  "line": 14,
                },
              },
              "operator": "++",
              "prefix": false,
              "start": 223,
              "type": "UpdateExpression",
            },
            "modifiers": [],
            "name": "click",
            "start": 212,
            "type": "EventHandler",
          },
        ],
        "children": [
          {
            "data": "
  Count is: ",
            "end": 246,
            "raw": "
  Count is: ",
            "start": 233,
            "type": "Text",
          },
          {
            "end": 255,
            "expression": Node {
              "end": 253,
              "loc": SourceLocation {
                "end": Position {
                  "column": 19,
                  "line": 15,
                },
                "start": Position {
                  "column": 14,
                  "line": 15,
                },
              },
              "name": "count",
              "start": 248,
              "type": "Identifier",
            },
            "start": 246,
            "type": "MustacheTag",
          },
          {
            "data": "
",
            "end": 256,
            "raw": "
",
            "start": 255,
            "type": "Text",
          },
        ],
        "end": 265,
        "name": "button",
        "start": 204,
        "type": "Element",
      },
    ],
    "end": 265,
    "start": 1,
    "type": "Fragment",
  },
  "instance": {
    "content": Node {
      "body": [
        Node {
          "end": 127,
          "loc": SourceLocation {
            "end": Position {
              "column": 34,
              "line": 6,
            },
            "start": Position {
              "column": 2,
              "line": 6,
            },
          },
          "source": Node {
            "end": 127,
            "loc": SourceLocation {
              "end": Position {
                "column": 34,
                "line": 6,
              },
              "start": Position {
                "column": 26,
                "line": 6,
              },
            },
            "raw": "'svelte'",
            "start": 119,
            "type": "Literal",
            "value": "svelte",
          },
          "specifiers": [
            Node {
              "end": 111,
              "imported": Node {
                "end": 111,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 18,
                    "line": 6,
                  },
                  "start": Position {
                    "column": 11,
                    "line": 6,
                  },
                },
                "name": "onMount",
                "start": 104,
                "type": "Identifier",
              },
              "loc": SourceLocation {
                "end": Position {
                  "column": 18,
                  "line": 6,
                },
                "start": Position {
                  "column": 11,
                  "line": 6,
                },
              },
              "local": Node {
                "end": 111,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 18,
                    "line": 6,
                  },
                  "start": Position {
                    "column": 11,
                    "line": 6,
                  },
                },
                "name": "onMount",
                "start": 104,
                "type": "Identifier",
              },
              "start": 104,
              "type": "ImportSpecifier",
            },
          ],
          "start": 95,
          "type": "ImportDeclaration",
        },
        Node {
          "declarations": [
            Node {
              "end": 144,
              "id": Node {
                "end": 140,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 11,
                    "line": 8,
                  },
                  "start": Position {
                    "column": 6,
                    "line": 8,
                  },
                },
                "name": "count",
                "start": 135,
                "type": "Identifier",
              },
              "init": Node {
                "end": 144,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 15,
                    "line": 8,
                  },
                  "start": Position {
                    "column": 14,
                    "line": 8,
                  },
                },
                "raw": "0",
                "start": 143,
                "type": "Literal",
                "value": 0,
              },
              "loc": SourceLocation {
                "end": Position {
                  "column": 15,
                  "line": 8,
                },
                "start": Position {
                  "column": 6,
                  "line": 8,
                },
              },
              "start": 135,
              "type": "VariableDeclarator",
            },
          ],
          "end": 144,
          "kind": "let",
          "loc": SourceLocation {
            "end": Position {
              "column": 15,
              "line": 8,
            },
            "start": Position {
              "column": 2,
              "line": 8,
            },
          },
          "start": 131,
          "type": "VariableDeclaration",
        },
        Node {
          "end": 193,
          "expression": Node {
            "arguments": [
              Node {
                "async": false,
                "body": Node {
                  "body": [
                    Node {
                      "end": 188,
                      "expression": Node {
                        "arguments": [
                          Node {
                            "end": 187,
                            "loc": SourceLocation {
                              "end": Position {
                                "column": 23,
                                "line": 11,
                              },
                              "start": Position {
                                "column": 16,
                                "line": 11,
                              },
                            },
                            "raw": "'mount'",
                            "start": 180,
                            "type": "Literal",
                            "value": "mount",
                          },
                        ],
                        "callee": Node {
                          "computed": false,
                          "end": 179,
                          "loc": SourceLocation {
                            "end": Position {
                              "column": 15,
                              "line": 11,
                            },
                            "start": Position {
                              "column": 4,
                              "line": 11,
                            },
                          },
                          "object": Node {
                            "end": 175,
                            "loc": SourceLocation {
                              "end": Position {
                                "column": 11,
                                "line": 11,
                              },
                              "start": Position {
                                "column": 4,
                                "line": 11,
                              },
                            },
                            "name": "console",
                            "start": 168,
                            "type": "Identifier",
                          },
                          "optional": false,
                          "property": Node {
                            "end": 179,
                            "loc": SourceLocation {
                              "end": Position {
                                "column": 15,
                                "line": 11,
                              },
                              "start": Position {
                                "column": 12,
                                "line": 11,
                              },
                            },
                            "name": "log",
                            "start": 176,
                            "type": "Identifier",
                          },
                          "start": 168,
                          "type": "MemberExpression",
                        },
                        "end": 188,
                        "loc": SourceLocation {
                          "end": Position {
                            "column": 24,
                            "line": 11,
                          },
                          "start": Position {
                            "column": 4,
                            "line": 11,
                          },
                        },
                        "optional": false,
                        "start": 168,
                        "type": "CallExpression",
                      },
                      "loc": SourceLocation {
                        "end": Position {
                          "column": 24,
                          "line": 11,
                        },
                        "start": Position {
                          "column": 4,
                          "line": 11,
                        },
                      },
                      "start": 168,
                      "type": "ExpressionStatement",
                    },
                  ],
                  "end": 192,
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 3,
                      "line": 12,
                    },
                    "start": Position {
                      "column": 16,
                      "line": 10,
                    },
                  },
                  "start": 162,
                  "type": "BlockStatement",
                },
                "end": 192,
                "expression": false,
                "generator": false,
                "id": null,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 3,
                    "line": 12,
                  },
                  "start": Position {
                    "column": 10,
                    "line": 10,
                  },
                },
                "params": [],
                "start": 156,
                "type": "ArrowFunctionExpression",
              },
            ],
            "callee": Node {
              "end": 155,
              "loc": SourceLocation {
                "end": Position {
                  "column": 9,
                  "line": 10,
                },
                "start": Position {
                  "column": 2,
                  "line": 10,
                },
              },
              "name": "onMount",
              "start": 148,
              "type": "Identifier",
            },
            "end": 193,
            "loc": SourceLocation {
              "end": Position {
                "column": 4,
                "line": 12,
              },
              "start": Position {
                "column": 2,
                "line": 10,
              },
            },
            "optional": false,
            "start": 148,
            "type": "CallExpression",
          },
          "loc": SourceLocation {
            "end": Position {
              "column": 4,
              "line": 12,
            },
            "start": Position {
              "column": 2,
              "line": 10,
            },
          },
          "start": 148,
          "type": "ExpressionStatement",
        },
      ],
      "end": 194,
      "loc": SourceLocation {
        "end": Position {
          "column": 0,
          "line": 13,
        },
        "start": Position {
          "column": 0,
          "line": 1,
        },
      },
      "sourceType": "module",
      "start": 92,
      "type": "Program",
    },
    "context": "default",
    "end": 203,
    "start": 74,
    "type": "Script",
  },
  "module": {
    "content": Node {
      "body": [
        Node {
          "declaration": Node {
            "declarations": [
              Node {
                "end": 63,
                "id": Node {
                  "end": 55,
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 18,
                      "line": 3,
                    },
                    "start": Position {
                      "column": 15,
                      "line": 3,
                    },
                  },
                  "name": "FOO",
                  "start": 52,
                  "type": "Identifier",
                },
                "init": Node {
                  "end": 63,
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 26,
                      "line": 3,
                    },
                    "start": Position {
                      "column": 21,
                      "line": 3,
                    },
                  },
                  "raw": "'BAR'",
                  "start": 58,
                  "type": "Literal",
                  "value": "BAR",
                },
                "loc": SourceLocation {
                  "end": Position {
                    "column": 26,
                    "line": 3,
                  },
                  "start": Position {
                    "column": 15,
                    "line": 3,
                  },
                },
                "start": 52,
                "type": "VariableDeclarator",
              },
            ],
            "end": 63,
            "kind": "const",
            "loc": SourceLocation {
              "end": Position {
                "column": 26,
                "line": 3,
              },
              "start": Position {
                "column": 9,
                "line": 3,
              },
            },
            "start": 46,
            "type": "VariableDeclaration",
          },
          "end": 63,
          "loc": SourceLocation {
            "end": Position {
              "column": 26,
              "line": 3,
            },
            "start": Position {
              "column": 2,
              "line": 3,
            },
          },
          "source": null,
          "specifiers": [],
          "start": 39,
          "type": "ExportNamedDeclaration",
        },
      ],
      "end": 64,
      "loc": SourceLocation {
        "end": Position {
          "column": 0,
          "line": 4,
        },
        "start": Position {
          "column": 0,
          "line": 1,
        },
      },
      "sourceType": "module",
      "start": 36,
      "type": "Program",
    },
    "context": "module",
    "end": 73,
    "start": 1,
    "type": "Script",
  },
}
`;

exports[`svelte compiler > preprocess 1`] = `
{
  "code": "
<script context="module" lang="ts">export const FOO = "BAR";
</script>
<script lang="ts">import { onMount } from "svelte";
let count = 0;
onMount(() => {
  console.log("mount");
});
</script>
<button on:click="{count++}">
  Count is: { count }
</button>
",
  "dependencies": [],
  "map": SourceMap {
    "mappings": ";mCACS,aAAM,MAAM;;kBAAnB,SAAS,eAAe;AAExB,IAAI,QAAQ;AAEZ,QAAQ,MAAM;AACZ,UAAQ,IAAI,OAAO;AACrB,CAAC",
    "names": [],
    "sourceRoot": undefined,
    "sources": [
      "<stdin>",
    ],
    "version": 3,
  },
  "toString": [Function],
}
`;
